### PR TITLE
Elektrify with highlevel API and code-generation

### PR DESCRIPTION
**Pull request by kodebach**
Changes:

- [x] lcdproc updated
- [x] lcdvc updated
- [x] lcdexec updated
- [x] LCDd (server core) updated
- [ ] drivers updated:
  - [x] bayrad
  - [x] cfontz
  - [x] cfontzpacket
  - [x] curses
  - [x] cwlnx
  - [ ] ea65
  - [ ] eyeboxone
  - [ ] futaba
  - [ ] g15
  - [x] glcd
  - [ ] glcdlib
  - [x] glk
  - [x] hd44780
  - [ ] icp_a106
  - [ ] imon
  - [ ] imonlcd
  - [ ] iowarrior
  - [ ] irman
  - [ ] joy
  - [x] lb216
  - [x] lcdm001
  - [ ] lcterm
  - [x] linux_input
  - [ ] lirc
  - [ ] lis
  - [ ] md8800
  - [ ] mdm166a
  - [ ] ms6931
  - [ ] mtc_s16209x
  - [x] mtxorb
  - [ ] mx5000
  - [ ] noritakevfd
  - [ ] olimex_mod_lcd1x9
  - [ ] picolcd
  - [x] pyramid
  - [ ] rawserial
  - [ ] sdeclcd
  - [ ] sed1330
  - [ ] sed1520
  - [ ] serialpos
  - [ ] serialvfd
  - [ ] shuttlevfd
  - [ ] sli
  - [ ] stv5730
  - [ ] svga
  - [ ] t6963
  - [x] text
  - [ ] tyan
  - [ ] ula200
  - [ ] vlsys_m428
  - [x] xosd
  - [ ] yard2lcd

---

Because of ElektraInitiative/libelektra#2634 the spec for drivers is essentially useless. To make the server work with the `curses` driver for example you need to do (at least) the following in addition to the normal mount + spec-mount:

```sh
kdb set "/sw/lcdproc/lcdd/#0/current/drivers/#0" curses
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/file" curses
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/background" cyan
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/foreground" blue
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/backlight" red
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/brightness" 1000
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/contrast" 1000
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/offbrightness" 0
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/reboot" 0
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/size" 20x4
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/topleftx" 7
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/toplefty" 7
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/useacs" 0
kdb set "/sw/lcdproc/lcdd/#0/current/curses/#0/drawborder" 1
```

### CODE REVIEW

**Comment by markus2330** [comment]
Is this draft PR a new feature? You tell me when I should look at this PR?

**Comment by kodebach** [comment]
> Is this draft PR a new feature?

I think it existed for some time, but it is pretty weird in some respects. e.g. CI systems seem to react differently to it, some run the build and other don't.

> You tell me when I should look at this PR?

Yes, I will tag you when it is ready. I found that I need an addition to the API to implement `lcdexec` the way I want to.

**Comment by markus2330** [comment]
> Yes, I will tag you when it is ready.

Note that I will not poll here, so please mention me.

**Comment by kodebach** [comment]
@markus2330 You can take a look at the code now. If you want to compile it, you need to use the Elektra version from ElektraInitiative/libelektra#2707.

Actually using the new version will be tedious however, because I still need to fix the `spec` plugin, so defaults for array elements work. So until that is done, I suggest you only look at the code. If you really want to try stuff, start by copying `curses` config from above and expect lots of "Key xyz not found" errors where you manually need to set the default.

My plan is to not update anymore drivers and work on my thesis instead. The remaining drivers should mostly be straight-forward to update. I will probably do it, once my thesis is finished. But since I can't even test most of the drivers that I updated already, I don't really see any reason to waste time.

The `readconf.py` needs an update too, so using that won't work either right now.

Also I noticed that having boolean keys with default value `true` doesn't work. I don't know whether that is something that can be fixed though, because I think we would need to transform the `default` metakey in the  `type` for that.

**Comment by markus2330** [comment]
> So until that is done, I suggest you only look at the code.

Ok!

> The remaining drivers should mostly be straight-forward to update.

Are there any important drivers missing? Maybe @Piankero can help out?

> The readconf.py needs an update too, so using that won't work either right now.

see #8 

> Also I noticed that having boolean keys with default value true doesn't work. I don't know whether that is something that can be fixed though, because I think we would need to transform the default metakey in the type for that.

I opened https://github.com/ElektraInitiative/libelektra/issues/2723

File: clients/lcdproc/disk.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #ifndef DISK_H
 #define DISK_H
```
Comment (markus2330) [inline clients/lcdproc/disk.h:3]: Any reasons you did not include elektragen.h everywhere where Elektra is needed?

(Otherwise it might lead to a compile error if header files are reordered.)

File: configure.ac [hunk 2]
```diff
@@ -76,6 +86,16 @@ AC_PROG_RANLIB
...
+dnl elektra
+ifdef([PKG_CHECK_MODULES],
+		[PKG_CHECK_MODULES([ELEKTRA], [elektra-codegen], [ELEKTRA_LIBDIR=`pkg-config --variable=libdir elektra-codegen`],)],
+		[AC_MSG_ERROR([pkg-config not (fully) installed, but needed to find Elektra])])
+AC_SUBST([ELEKTRA_CFLAGS])
+AC_SUBST([ELEKTRA_LIBS])
+
+CFLAGS="$ELEKTRA_CFLAGS $CFLAGS"
+LIBS="$ELEKTRA_LIBS $LIBS -Wl,-rpath $ELEKTRA_LIBDIR"
+
```
Comment (markus2330) [inline configure.ac:90]: (does not belong to this line) Why is  lcdd_spec.ini and lcdexec_spec.ini binary?

File: server/drivers/CFontzPacket.c [hunk 3]
```diff
@@ -149,38 +148,35 @@ CFontzPacket_init (Driver *drvthis)
 	EmptyKeyRing(&keyring);
 	EmptyReceiveBuffer(&receivebuffer);
 
-	/* Read config file */
+	/* Read config */
+	CFontzPacketDriverConfig config;
+	elektraGet2V(elektra, &config, ELEKTRA_TAG_CFONTZPACKET, drvthis->index);
```
Comment (markus2330) [inline server/drivers/CFontzPacket.c]: 2V?

File: server/drivers/MD8800.c [hunk 4]
```diff
@@ -3,7 +3,7 @@
  */
 
 /*
-    Copyright (C) 2006 Stefan Herdler in collaboration with Martin M�ller.
+    Copyright (C) 2006 Stefan Herdler in collaboration with Martin M�ller.
```
Comment (markus2330) [inline server/drivers/MD8800.c]: Is this change on purpose? GitHub now renders the character wrongly (and did correctly before).

File: server/drivers/MD8800.h [hunk 5]
```diff
@@ -1,6 +1,6 @@
 /*  This is the LCDproc driver for the VFD of the Medion MD8800 PC
 
-    Copyright (C) 2006 Stefan Herdler in collaboration with Martin M�ller.
+    Copyright (C) 2006 Stefan Herdler in collaboration with Martin M�ller.
```
Comment (markus2330) [inline server/drivers/MD8800.h]: Also here.

File: server/drivers/MtxOrb.c [hunk 6]
```diff
@@ -362,36 +335,65 @@ MtxOrb_init (Driver *drvthis)
...
-			s = drvthis->config_get_string(drvthis->name, buf, 0, NULL);
-
-			/* Was a key specified in the config file ? */
-			if (s != NULL) {
-				p->keys++;
-				p->keymap[i] = strdup(s);
-				report(RPT_INFO, "%s: Key '%c' mapped to \"%s\"",
-					drvthis->name, i+'A', s );
-			}
+		if(strlen(config.keymapA) > 0) {
```
Comment (markus2330) [inline server/drivers/MtxOrb.c]: Put this to a function? (quite often duplicated)

File: server/drivers/curses_drv.c [hunk 7]
```diff
@@ -705,25 +699,27 @@ curses_wborder (Driver *drvthis)
...
-		return COLOR_MAGENTA;
-	else if (strcasecmp(colorname, "cyan") == 0)
-		return COLOR_CYAN;
-	else if (strcasecmp(colorname, "white") == 0)
-		return COLOR_WHITE;
-
-	return default_color;
+map_color_to_curses (CursesColor color) {
+	switch(color) {
+		case CURSES_COLOR_RED:
```
Comment (markus2330) [inline server/drivers/curses_drv.c:704]: Why does the code not directly use the generated enum?

File: server/drivers/pylcd.c [hunk 8]
```diff
@@ -469,22 +472,22 @@ pyramid_flush(Driver *drvthis)
 	 */
 	for (i = 1; i < 33; i++) {
 	    switch ((unsigned char)mesg[i]) {
-	    case 0xe4:		/* � */
+	    case 0xe4:		/* � */
```
Comment (markus2330) [inline server/drivers/pylcd.c]: Is this change on purpose?

File: server/drivers/xosdlib_drv.c [hunk 9]
```diff
@@ -85,74 +80,53 @@ xosdlib_drv_init (Driver *drvthis)
...
 	/* Which backlight-off "brightness" */
-	tmp = drvthis->config_get_int(drvthis->name, "OffBrightness", 0, DEFAULT_OFFBRIGHTNESS);
-	debug(RPT_INFO, "%s: OffBrightness (in config) is '%d'", __FUNCTION__, tmp);
-	if ((tmp < 0) || (tmp > 1000)) {
-		report(RPT_WARNING, "%s: OffBrightness must be between 0 and 1000; using default %d",
-			drvthis->name, DEFAULT_OFFBRIGHTNESS);
-		tmp = DEFAULT_OFFBRIGHTNESS;
-	}
-	p->offbrightness = tmp;
+	debug(RPT_INFO, "%s: OffBrightness (in config) is '%d'", __FUNCTION__, config.offbrightness);
```
Comment (markus2330) [inline server/drivers/xosdlib_drv.c:121]: Are debug statements like these everywhere?

File: clients/lcdproc/main.c [hunk 10]
```diff
@@ -36,8 +36,8 @@
 #include "mode.h"
 #include "shared/sockets.h"
 #include "shared/report.h"
-#include "shared/configfile.h"
```
Comment (markus2330) [inline clients/lcdproc/main.c:39]: Did you remove shared/configfile.c in this PR?

File: clients/lcdproc/main.c [hunk 11]
```diff
@@ -36,8 +36,8 @@
...
-#include "getopt.h"		/* This is our local getopt.h! */
```
Comment (markus2330) [inline clients/lcdproc/main.c:40]: Did you remove getopt.h in this PR?

File: clients/lcdexec/lcdexec.c [hunk 12]
```diff
@@ -373,13 +287,145 @@ static int connect_and_setup(void)
...
 
+/**
+ * returns the parameter with the given id, if it is contained in the given command
+ * returns NULL otherwise
+ */
+static CommandParameter * find_parameter(Command * command, int id, CommandParameterType * type)
+{
+	for (kdb_long_long_t i = 0; i < command->parametersSize; i++)
+	{
+		switch (command->parameterTypes[i])
```
Comment (markus2330) [inline clients/lcdexec/lcdexec.c:302]: Can we use the code-generator to avoid mappings like these?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great work! Unfortunately, quite some mappings are still needed or were even introduced. I would like to read more about the necessity of these mappings (maybe as future work how to avoid them).

File: clients/lcdproc/disk.h (hunk 1 above)
Comment (kodebach) [reply clients/lcdproc/disk.h:3]: You are right, it should be included here.

File: configure.ac (hunk 2 above)
Comment (kodebach) [reply configure.ac:90]: Those files shouldn't even exist. I need to look into it, but mounting the specification creates those files. I think is a result of using `specload` and `noresolver` together.

(The files should either be empty or contain only the `quickdump` magicnumber.)

File: server/drivers/CFontzPacket.c (hunk 3 above)
Comment (kodebach) [reply server/drivers/CFontzPacket.c]: `V` is for the vararg version.

`2` is just because I didn't know what to call the version of `elektraGet` that uses a result pointer instead of returning a value. This is version of `elektraGet` only exists for structs and doesn't malloc.

File: server/drivers/MD8800.c (hunk 4 above)
Comment (kodebach) [reply server/drivers/MD8800.c]: No, it seems my editor messed up the file encoding (possible during a regex replace). I will fix all these issues.

File: server/drivers/curses_drv.c (hunk 7 above)
Comment (kodebach) [reply server/drivers/curses_drv.c:704]: The used values (`COLOR_RED`, etc) come from the curses library.

File: server/drivers/xosdlib_drv.c (hunk 9 above)
Comment (kodebach) [reply server/drivers/xosdlib_drv.c:121]: The whole LCDproc code base is very inconsistent (with formatting, logging, etc). The recommendation seems to be to just keep going in the style of the original author. 

I neither added nor deleted debug or report statements AFAIK.

File: clients/lcdproc/main.c (hunk 10 above)
Comment (kodebach) [reply clients/lcdproc/main.c:39]: Right now, it is still there, but it should be unused and will be removed.

File: clients/lcdproc/main.c (hunk 11 above)
Comment (kodebach) [reply clients/lcdproc/main.c:40]: It is still there, but I think I will remove it. I don't think there is any reason to use the local (13 year old) version.

File: clients/lcdexec/lcdexec.c (hunk 12 above)
Comment (kodebach) [reply clients/lcdexec/lcdexec.c:302]: What exactly do you mean? How would you solve this differently (without thinking whether or not that is possible through code generation)?

File: server/drivers/CFontzPacket.c (hunk 3 above)
Comment (markus2330) [reply server/drivers/CFontzPacket.c]: > 2 is just because I didn't know what to call the version of elektraGet that uses a result pointer instead of returning a value

Please always create issues for name discussions in such situations. If such names creep into the API you never get rid of them again.

Please do so now for API names where you have doubts that they are perfect.

File: server/drivers/xosdlib_drv.c (hunk 9 above)
Comment (markus2330) [reply server/drivers/xosdlib_drv.c:121]: > The recommendation seems to be to just keep going in the style of the original author.

Yes, that makes sense. It is too much effort to make this consistent now.

File: clients/lcdproc/main.c (hunk 11 above)
Comment (markus2330) [reply clients/lcdproc/main.c:40]: Yes, please remove all these parts which are not needed. Reducing amount of code is one of the main reasons for this whole endeavor.

File: clients/lcdexec/lcdexec.c (hunk 12 above)
Comment (markus2330) [reply clients/lcdexec/lcdexec.c:302]: > What exactly do you mean?

I only wonder why this mapping is needed. It seems like it was not needed before?

Comment (kodebach) [reply clients/lcdexec/lcdexec.c:302]: It was needed but differently.

Before all menu entries (whether they were menus, commands or parameters) were represented by the same struct. Now we use separate types for different types of menu entries.

Before it was something like:
```c
struct {
    int id;
    union {
       // other stuff
    };
};
```

Now we use something like:
```c
union {
    struct {
       int id;
       // other stuff
    },
   struct {
       int id;
       // other stuff
    };
};
```

Of course this pseudo-code is not correct, but the point is the `id` is inside the union elements, so we need the switch, because it is not guaranteed that the `id` is always stored at the same offset.

The reason for the switch is that the first version would be very difficult to support using code-generation. Even now, we iterate over the menu and add the unique `id`s after Elektra has loaded the whole tree.

**Comment by kodebach** [comment]
`readconf.py` should now be able to convert old client configurations into the new versions. The conversion of `LCDd.conf` files is not supported yet. However, I think `lcdexec` configurations are the only ones that are really in need of automatic conversion, since those would be rather tedious to update manually.

**Comment by kodebach** [comment]
There is now some basic documentation about how to use this PR. I will update it further once I finish the documentation on the Elektra side.

Also, for now the documentation is in separate files, since that was quicker than updating the old ones.

**Comment by markus2330** [comment]
@kodebach can you please merge this to master?

If possible, please also rebase our master (of https://github.com/ElektraInitiative/lcdproc) to the current master of LCDproc.

**Comment by markus2330** [comment]
Thank you! So @haraldg can now test the master.
